template/interpolate/azure/keyvault/keyvault.go

```diff
@@ -0,0 +1,162 @@
+package keyvault
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+
+	keyvaultmgmt "github.com/Azure/azure-sdk-for-go/services/keyvault/2016-10-01/keyvault"
+	"github.com/Azure/azure-sdk-for-go/services/keyvault/mgmt/2018-02-14/keyvault"
+	"github.com/hashicorp/go-azure-helpers/authentication"
+	"github.com/hashicorp/go-azure-helpers/sender"
+)
+
+type Client struct {
+	VaultsClient     *keyvault.VaultsClient
+	ManagementClient *keyvaultmgmt.BaseClient
+}
+
+func New() (*Client, error) {
+	c := &Client{}
+
+	vaultsClient, managementClient, err := newVaultsClients()
+	if err != nil {
+		return nil, err
+	}
+
+	c.VaultsClient = vaultsClient
+	c.ManagementClient = managementClient
+
+	return c, nil
+}
+
+func (c *Client) GetKeyVaultId(resourceGroup string, keyVaultName string) (string, error) {
+	resp, err := c.VaultsClient.Get(context.Background(), resourceGroup, keyVaultName)
+	if err != nil {
+		return "", err
+	}
+
+	return *resp.ID, nil
+}
+
+func (c *Client) GetSecret(keyVaultId string, key string) (string, error) {
+	keyVaultBaseUri, err := GetKeyVaultBaseUrlFromID(context.Background(), c.VaultsClient, keyVaultId)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := c.ManagementClient.GetSecret(context.Background(), keyVaultBaseUri, key, "")
+	if err != nil {
+		return "", err
+	}
+
+	return *resp.Value, nil
+}
+
+func newVaultsClients() (*keyvault.VaultsClient, *keyvaultmgmt.BaseClient, error) {
+	builder := &authentication.Builder{
+		SupportsClientCertAuth:   true,
+		SupportsClientSecretAuth: true,
```

Comment:
From reading the code (e.g. [here](https://github.com/hashicorp/go-azure-helpers/blob/f9459454de0c5ef34d233dc3502ae273875ed723/authentication/auth_method_client_secret.go#L31)), it seems like `SupportsClientCertAuth` and `SupportsClientSecretAuth` need additional properties set to actually do something?

I see further down that these funcs are to be used in the variables section only, so I understand that builder Azure auth config is not available here. Not sure how to solve this elegantly?

We need to at least document this very clearly.

Reply:
Let me have a look at this as I am not an expert in all Azure auth methods and will comment here